src/main/scala/reporting/Converter.scala

```diff
@@ -0,0 +1,511 @@
+) extends ExtractedModelEntry {
+  override def toString: String = {
+    val buf = fields
+      .map(x =>
+        s"\t${x._1}(perm: ${x._2._2.map(_.toString).getOrElse("?")}) <- ${x._2._1.toString.split("\n").mkString("\n\t")}\n"
+      )
+      .mkString("")
+    s"Ref ($name) {\n$buf}"
+  }
+}
+case class NullRefEntry(name: String) extends ExtractedModelEntry {
+  override def toString = s"Null($name)"
+}
+case class RecursiveRefEntry(name: String) extends ExtractedModelEntry {
+  override def toString = s"recursive reference to $name"
+}
+case class VarEntry(name: String, sort: Sort) extends ExtractedModelEntry {
+  override def toString: String = name
+}
+
+case class OtherEntry(value: String, problem: String = "")
+    extends ExtractedModelEntry {
+  override def toString = s"$value [$problem]"
+}
+case class SeqEntry(name: String, values: List[ExtractedModelEntry])
+    extends ExtractedModelEntry {
+  override def toString = s"($name): [${values.map(_.toString).mkString(", ")}]"
+}
+
+case class UnprocessedModelEntry(entry: ValueEntry)
+    extends ExtractedModelEntry {
+  override def toString = s"$entry"
+}
+
+// processed Heap representation:
+sealed trait HeapEntry
+
+case class ExtractedHeap(entries: List[HeapEntry])
+
+case class PredHeapEntry(
+    name: String,
+    args: Seq[ExtractedModelEntry]
+) extends HeapEntry {
+  override def toString = s"$name(${args.mkString(", ")})"
+}
+case class FieldHeapEntry(
+    recv: VarEntry,
+    field: String,
+    perm: Option[Rational],
+    sort: Sort,
+    entry: ExtractedModelEntry
+) extends HeapEntry {
+  override def toString = s"$recv.$field"
+}
+
+case class UnresolvedHeapEntry(chunk: Chunk, reason: String) extends HeapEntry {
+  override def toString = s"$chunk (not further processed because: $reason)"
+}
+
+object Converter {
+
+  def getFunctionValue(
+      model: Model,
+      fname: String,
+      args: Seq[ValueEntry],
+      toSort: Sort
+  ): ExtractedModelEntry = {
+    val entry: Option[ModelEntry] = model.entries.get(fname)
+    entry match {
+      case Some(MapEntry(m, els)) =>
+        getConstantEntry(toSort, m.getOrElse(args, els))
+      case Some(m) => getConstantEntry(toSort, m)
+      case None    => OtherEntry("${fname}", "function not found in model")
+    }
+  }
+
+  def translateSort(s: Sort): String = {
+    s match {
+      case sorts.Set(els: Sort) => s"Set<${translateSort(els)}>"
+      case sorts.Ref            => "$Ref"
+      case sorts.Snap           => "$Snap"
+      case sorts.Perm           => "$Perm"
+      case sorts.Seq(els: Sort) => s"Seq<${translateSort(els)}>"
+      case _                    => s.toString
+    }
+  }
+
+  def getConstantEntry(s: Sort, m: ModelEntry): ExtractedModelEntry = {
+    s match {
+      case sorts.Ref => VarEntry(m.toString, sorts.Ref)
+      case sorts.Int =>
+        m match {
+          case ConstantEntry(x)             => LitIntEntry(BigInt(x))
+          case ApplicationEntry(name, args) =>
```

Comment:
Can this match not be done as `ApplicationEntry("-", [LitIntEntry(n)]) => LitIntEntry(-n)` (or similar?) Is `negate` used anywhere else?

Reply:
You are right, this is not really necessary and no, negate is not used anywhere else.